src/ingest_validation_tools/upload.py

```diff
@@ -106,124 +85,140 @@ def __init__(
+            tsvs = {
                 Path(path).name: {
                     "Schema": sv.table_schema,
                     "Metadata schema version": sv.version,
                     "Directory schema version": sv.dir_schema,
                 }
                 for path, sv in self.effective_tsv_paths.items()
             }
-            return {
-                "Time": datetime.now(),
-                "Git version": git_version,
-                "Directory": str(self.directory_path),
-                "TSVs": effective_tsvs,
-            }
         except PreflightError as e:
-            self.errors["Preflight"] = e
-            return self.errors
+            self.errors.preflight.append(str(e))
+            return
 
-    def get_errors(self, **kwargs) -> dict:
-        # This creates a deeply nested dict.
-        # Keys are present only if there is actually an error to report.
-        # plugin_kwargs are passed to the plugin validators.
+        return InfoDict(
+            time=datetime.now(),
+            git=git_version,
+            dir=str(self.directory_path),
+            tsvs=tsvs,
+        )
 
+    def get_errors(self, **kwargs) -> ErrorDict:
+        """
+        This creates an ErrorDict object
+        When converted using ErrorDict.as_dict(), keys are
+        present only if there is actually an error to report.
+        """
+        # plugin_kwargs are passed to the plugin validators via extra_parameters
         kwargs.update(self.extra_parameters)
+
+        # Return if PreflightErrors found
         if self.errors:
             return self.errors
 
         if not self.effective_tsv_paths:
-            return {"Missing": "There are no effective TSVs."}
-
-        errors = {}
-        try:
-            upload_errors = self._check_upload()
-            if upload_errors:
-                errors["Upload Errors"] = upload_errors
-
-            validation_errors = self.validation_routine()
-            if validation_errors:
-                errors["Metadata TSV Validation Errors"] = validation_errors
-
-            reference_errors = self._get_reference_errors()
-            if reference_errors:
-                errors["Reference Errors"] = reference_errors
-
-            if errors and not self.run_plugins:
-                raise ErrorDictException(
-                    "Skipping plugins validation: errors in upload metadata or dir structure."
-                )
+            self.errors.preflight.append("There are no effective TSVs.")
+            return self.errors
 
-            elif self.run_plugins:
-                logging.info("Running plugin validation...")
+        # Collect errors
+        self._get_local_tsv_errors()
+        self._get_directory_errors()
+        self.validation_routine()
+        self._get_reference_errors()
+
+        # Plugin error checking is costly, by default this bails
+        # if other errors have been found already
+        if self.errors and not self.run_plugins:
```

Comment:
To answer your first question, yes. I'm thinking now that this isn't necessary, because the default behavior you describe is sensible. I'll test and make sure.

Reply:
Seems to work fine!